sdk/python/lib/pulumi/_utils.py

```diff
@@ -119,3 +120,87 @@ def lazy_import(fullname):
     sys.modules[fullname] = module
     loader.exec_module(module)
     return module
+
+
+def contextproperty(_fn=None, *, default: typing.Optional[typing.Any] = None):
+    def inner(func: typing.Callable):
+        new_property = ContextProperty(
+            name=func.__qualname__, doc=func.__doc__, default=default
+        )
+        inner.__doc__ = new_property.__doc__
+        return new_property
+
+    if _fn is None:
+        return inner
+    return inner(_fn)
+
+
+class ContextProperty:
```

Comment:
doc comments on what this is for would be good, it's not going to be immediatly clear why we're not just using contextvars  directly

Reply:
Sure!  I'll definitely document this.

Reply:
For completeness, the added doctests _do_ pass should we ever want to start supporting them:
![image](https://user-images.githubusercontent.com/4128006/186154192-2b203f18-5bb9-4124-8d57-8c3a5342c388.png)